leapp/utils/lock.py

```diff
@@ -0,0 +1,96 @@
+import os
+import fcntl
+import logging
+
+from leapp.config import get_config
+from leapp.exceptions import ProcessLockError
+
+
+def leapp_lock(lockfile=None):
+    return ProcessLock(lockfile=lockfile)
+
+
+def _acquire_lock(fd):
+    try:
+        fcntl.flock(fd, fcntl.LOCK_EX | fcntl.LOCK_NB)
+        return True
+    except OSError:
+        return False
+
+
+def _read_pid(fd):
+    return os.read(fd, 20)
+
+
+def _write_pid(fd, pid):
+    os.write(fd, str(pid).encode('utf-8'))
+
+
+def _validate_pid(old_pid):
+    try:
+        old_pid = int(old_pid)
+    except ValueError:
+        return False
+
+    return old_pid
+
+
+def _clear_lock(fd):
+    os.lseek(fd, 0, os.SEEK_SET)
+    os.ftruncate(fd, 0)
+
+
+class ProcessLock(object):
+
+    def __init__(self, lockfile=None):
+        self.log = logging.getLogger('leapp.utils.lock')
+        self.lockfile = lockfile if lockfile else get_config().get('lock', 'path')
+
+        self.fd = None
+
+    def _get_pid_from_lockfile(self):
+        running_pid = _read_pid(self.fd)
+        self.log.debug("_get_pid_from_lockfile: running_pid=%s", running_pid)
+
+        if len(running_pid) == 0:
+            # No pid in lockfile
+            return
+
+        running_pid = _validate_pid(running_pid)
+        if not running_pid:
+            return
+
+        return running_pid
+
+    def _try_lock(self, pid):
+        if not _acquire_lock(self.fd):
+            running_pid = self._get_pid_from_lockfile()
+
+            process_msg = ''
+            if running_pid:
+                process_msg = ' by process with PID {}'.format(running_pid)
+
+            msg = ('Leapp is currently locked{} and cannot be started.\n'
+                   'Please ensure no other instance of leapp is running and try again.').format(process_msg)
+            raise ProcessLockError(msg)
+
+        try:
+            _clear_lock(self.fd)
+            _write_pid(self.fd, pid)
+        except OSError:
+            raise ProcessLockError('Could not write PID to lockfile.')
```

Comment:
Do we want this to stop execution or to log and continue?  Now that the PID in the lockfile is only informational, I could see it either way.

Reply:
Hmm. Good question. thinking about it, in case it's hyphotethical situation, most likely doesn't matter whether we raise it or whether we continue. But if this happens, it's question why it should happen and whether it could have another impact on the execution - not just speaking about having or not the content in the file, but speaking about the root cause, whether the same issue could have another impact on the rest of the leapp execution that could lead to another set of errors later.